### ISSUE TITLE
[Bug]: terraform.tfstate has not actual state after terraform apply

### ISSUE THREAD

**Issue by EugenKon** [hashicorp/terraform-provider-aws#37807]
### Terraform Core Version

v1.8.4

### AWS Provider Version

v4.67.0

### Affected Resource(s)

* aws_lb

### Expected Behavior

After 
```
# Change subnets:
terraform plan -target aws_lb.public -out alb
terraform apply alb
# Change other subnets:
terraform plan -target aws_lb.public -out alb
terraform apply alb
```

`subnet_mapping` should be in actual state, eg. all new subnets should be there.

### Actual Behavior

After the first change was applied:
```diff
resource "aws_lb" "public" {
  name         = "public-WWW"

  subnets = [
    aws_subnet.a_prd_plntr.id,
    aws_subnet.b_prd_plntr.id,
-    aws_subnet.c_prd_plntr.id,
-    aws_subnet.d_prd_plntr.id,
+    aws_subnet.c_public_prd_plntr.id,
+    aws_subnet.d_public_prd_plntr.id,
  ]
  ...
}
```
I got correct state.

But after the second change was applied:
```diff
resource "aws_lb" "public" {
  name         = "public-WWW"

  subnets = [
-    aws_subnet.a_prd_plntr.id,
-    aws_subnet.b_prd_plntr.id,
+    aws_subnet.a_public_prd_plntr.id,
+    aws_subnet.b_public_prd_plntr.id,
    aws_subnet.c_public_prd_plntr.id,
    aws_subnet.d_public_prd_plntr.id,
  ]
  ...
}
```
The new TF state does not reflect the actual state.
![2024-06-03_08-24](https://github.com/hashicorp/terraform-provider-aws/assets/139780495/9ffb71ed-4c91-4f4c-a907-8fd65fc1a7ec)

![2024-06-03_08-25_1](https://github.com/hashicorp/terraform-provider-aws/assets/139780495/85e82df6-a999-42e5-b78f-1dde88036ef6)

![2024-06-03_08-25](https://github.com/hashicorp/terraform-provider-aws/assets/139780495/2bb4186d-4e98-4529-9925-2e4fe600b7e1)

![2024-06-03_08-31](https://github.com/hashicorp/terraform-provider-aws/assets/139780495/934afd87-28fc-41ab-8a95-3dc0afad938f)

It was fixed only by running:
```
terraform plan -target aws_lb.public -out alb -refresh-only
terraform apply -refresh-only alb
```
![2024-06-03_09-13](https://github.com/hashicorp/terraform-provider-aws/assets/139780495/12ad0bda-f94c-461b-9ff4-7657c973f8e9)

The usual:
```
terraform plan -target aws_lb.public -out alb
```
does not see the changes:
![image](https://github.com/hashicorp/terraform-provider-aws/assets/139780495/32e9f3b8-2465-472b-808d-6c37242c5683)

### Terraform Configuration Files

See **actual behaviour** section.

### Steps to Reproduce

1. Change two subnets for ALB
2. Apply changes
3. Change another two subnets for ALB
4. Apply changes

Details are at **actual behaviour** section.

### Would you like to implement a fix?

None

**Comment by justinretzolk** [comment]
Hey @EugenKon 👋 Thank you for taking the time to raise this! We'll need a sample Terraform configuration that can be used to reproduce this in order to look into it. Are you able to provide that and, if possible, debug logs (redacted as needed)?

**Comment by EugenKon** [comment]
@justinretzolk Yes, I can send the full configuration. Can you provide the email address? Unfortunately the default one for private post does not work: https://github.com/hashicorp/nomad/issues/23168

**Comment by EugenKon** [comment]
@justinretzolk Do I miss something? This is the link not to "GPG public key", but to some Chat application. I am not sure how to use that chat application to encrypt files.

**Comment by justinretzolk** [comment]
@EugenKon Near the top of the page, the key's fingerprint is shown. If you click on it, a modal appears that has the public key. Thanks for following up on that -- something to keep in mind to add to the form as well.

**Comment by justinretzolk** [comment]
Hey @EugenKon 👋 I wanted to follow up on this one. I know on another issue, we discussed how the `-target` flag was the culprit for some aspects of Terraform not behaving as you'd expected; did this seem to be a similar situation?

**Comment by EugenKon** [comment]
I faced a lot of issues with `-target`. It feels like it was not implemented well and has a lot of underwater rocks. All reported issues regarding `-target` should be summarized and code refactored. This is my opinion.

All reported issues are separate issues with different scenarios. Though they are all similar: all the times when `-target` was used I had issues with tfstate.

**Comment by justinretzolk** [comment]
That makes a lot of sense, thank you for the follow up. As far as refactoring / changing how `-target` works, that would need to be done at the Terraform Core level. In the meantime, the [documentation for `-target`](https://developer.hashicorp.com/terraform/cli/commands/plan#resource-targeting) mentions that it really should only be used in exceptional circumstances, rather than day-to-day operations. 

> This targeting capability is provided for exceptional circumstances, such as recovering from mistakes or working around Terraform limitations. It is not recommended to use -target for routine operations, since this can lead to undetected configuration drift and confusion about how the true state of resources relates to configuration.

With that in mind, since there's no action for the AWS Provider team to take on this issue at this time, I'm going to close it out. If you experience any further unexpected behavior, please do let us know!

**Comment by EugenKon** [comment]
@justinretzolk That is really useful during migration tasks, or when not all resources are managed via terraform or when people do a lot of changes directly and do not honour TF configuration.
This is why I was forced to use `-target` so often =(.
